w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,259 @@
+                i.set_url(ptest_url)
+                self.kb_append(self, 'hpp', i)
+                return
+
+
+
+        if self.precedence == UNK:
+            om.out.debug(u'Unable to determine precedence for HPP detection.'
+                         u'Presumably not vulnerable to HPP.')
+
+            return
+#        print "Attack!!", self.precedence
+        om.out.debug(u"Detected precedence is %s."
+                     u"\nTrying pollution of the reflected parameters: %s."
+                     % (self.precedence, rparams.keys()))
+        #bulding payloads
+        payloads = []
+        vuln_params=[]
+        for k,v in rparams.items():
+            payloads.append(REFLECT_TOKEN+'&'+v+'='+FORGE_TOKEN)
+        vtest_mutants = create_mutants(freq, payloads,
+                                       fuzzable_param_list=rparams.keys(),
+                                       mutant_tuple=(QSMutant,))
+        for m in vtest_mutants:
+            vtest_resp = self._uri_opener.send_mutant(m)
+            try:
+                vtest_dp = parser_cache.dpc.get_document_parser_for(vtest_resp)
+            except BaseFrameworkException:
+                return
+            r =  self.find_reflection( vtest_dp, FORGE_TOKEN)
+            if r:
+                vuln_params.append(r)
+                msg=u"value of parameter %s can be forged with " \
+                    u"HPP attack using URL: %s, precedence is %s"
+                msg = msg % (r, m.get_uri(), self.precedence)
+                v=Vuln.from_mutant(u'HPP', desc=msg, mutant = m,
+                                   plugin_name='hpp',response_ids=[],
+                                   severity=severity.MEDIUM)
+                v.set_url(self._orig_uri)
+                v[HPPInfoSet.ITAG] = m.get_uri()
+                # debug_msg = debug_msg % (rid, freq.get_uri())
+                self.kb_append_uniq_group(self, 'hpp', v,
+                                          group_klass=HPPInfoSet)
+
+    def find_reflection(self, parser, pattern1, pattern2 = ''):
+        #finds reflection of a string or two concatenated
+        # strings in web page links or input parameters
+
+        result = False
+        #find reflection in links
+        for ref in parser.get_references():
+            if ref == []:
+                continue
+            qs = ref[0].get_querystring()
+            for p in qs:
+                if self.precedence == LAST and len(qs[p]) == 2:
+                  v =  qs[p][1]
+                else:
+                  v =  qs[p][0]
+                #looking for single reflection if only one pattern was entered
+                if v == pattern1 and not pattern2:
+                    result = p
+                #or for concatenation of both were
+                if v[:len(pattern1)] == pattern1 and \
+                                v[-len(pattern2):] == pattern2:
+                    result = p, ref[0]
+
+        return result
+
+
+    def strip_trash(self, doc):
```

Comment:
Hmmm, there might me dynamic content in HTML comments, true. But the probabilities are slim <5% ? <1% ?

There might also be dynamic content outside of HTML comments affecting the fuzzy_equal function, which has a much greater probability (compared with the previous) and we're not doing anything.

Also, what if the change in the HTML comment is actually related with the change in QS params you sent; and by removing the comments you're not finding it?

I would remove the strip_trash